front/pages/[user]/a/[sId]/execute/index.jsx

```diff
@@ -0,0 +1,403 @@
+import {
+  CheckCircleIcon,
+  ChevronDownIcon,
+  ChevronRightIcon,
+  PlayCircleIcon,
+} from "@heroicons/react/20/solid";
+import { unstable_getServerSession } from "next-auth/next";
+import { useSession } from "next-auth/react";
+import { useState, useEffect, useRef } from "react";
+import TextareaAutosize from "react-textarea-autosize";
+import { SSE } from "sse.js";
+import {
+  ArrayViewer,
+  ObjectViewer,
+  StringViewer,
+} from "../../../../../components/app/blocks/Output";
+import MainTab from "../../../../../components/app/MainTab";
+import AppLayout from "../../../../../components/AppLayout";
+import { ActionButton, Button } from "../../../../../components/Button";
+import { extractConfig } from "../../../../../lib/config";
+import { checkDatasetData } from "../../../../../lib/datasets";
+import { classNames } from "../../../../../lib/utils";
+
+const { URL, GA_TRACKING_ID = null } = process.env;
+
+function preProcessOutput(output) {
+  if (Array.isArray(output) && output.length === 1) {
+    return preProcessOutput(output[0]);
+  }
+  if (output.value && output.error === null) {
+    return preProcessOutput(output.value);
+  }
+  if (output.error) {
+    return preProcessOutput(output.error);
+  }
+  if (output.completion?.text) {
+    return output.completion.text;
+  }
+  return output;
+}
+
+function Spinner() {
```

Comment:
This is also duplicated code, need to move it in some other place

Reply:
Also this probably re-runs anytime an output is added so not very optimal

Reply:
Yeah that’d be nice to dry it 👍

Reply:
Moved it into `components` and replaced it in `Block.jsx` as well